### PR TITLE
Allow ADR plugin to use sites other than GitHub

### PR DESCRIPTION
**Pull request by Robunnin**
## Hey, I just made a Pull Request!

The adr plugin has been modified to work with sites other than GitHub. 

I accomplished this using UrlReaders. I have created the concept of an "AdrFileFetcher," which is an interface that has two hooks- one for getting the contents of a file at a provided url, and another for getting a listing of files at a provided url. I have added two AdrFileFetchers- octokitAdrFileFetcher and urlReaderAdrFileFetcher.

By default, it still uses octokit and will work only with GitHub like before (in case anyone prefers to use it this way, for instance because of rate limiting). To use the UrlReaders implementation, I have added an optional prop called "adrFileFetcher" on EntityAdrContent and AdrReader. If you assign that prop to urlReaderAdrFileFetcher (or any other AdrFileFetcher) it will switch to using the provided fetcher. 

To use UrlReaders, I have expanded the adr-backend plugin. I have added two endpoints: getAdrFilesAtUrl and readAdrFileAtUrl that the adr plugin uses to leverage the UrlReaders. They both take in a url through a query parameter and either return a listing of files at the provided url or return the contents of the file at the provided url. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by jhaals — CHANGES_REQUESTED** [review]
Thanks for the suggestion!

I'll let @kuangp take a look at this as he authored the plugin itself.
Make sure that CI errors are addressed and that API reports are generated with `yarn build:api-reports`

**Comment by djamaile** [comment]
hmm weird @kuangp didn't get pinged for this as he is in the `CODEOWNERS` file

**Comment by jhaals** [comment]
> hmm weird @kuangp didn't get pinged for this as he is in the `CODEOWNERS` file
Github does not support codeowners with read access. Will be fixed if we realise #15317

**Comment by Robunnin** [comment]
> Thanks for the suggestion!
> 
> I'll let @kuangp take a look at this as he authored the plugin itself. Make sure that CI errors are addressed and that API reports are generated with `yarn build:api-reports`

I have addressed the CI errors and generated the api reports. All checks are passing right now.

File: plugins/adr/src/hooks/adrFileFetcher.ts
Comment (jhaals) [inline plugins/adr/src/hooks/adrFileFetcher.ts:95]: Sorry but I don't understand the reason for having a fetcher with two methods when this previously just had one.

File: plugins/adr/src/components/EntityAdrContent/EntityAdrContent.tsx
Comment (jhaals) [inline plugins/adr/src/components/EntityAdrContent/EntityAdrContent.tsx:152]: Wouldn't this be problematic if you had URLs from multiple sources? I'm thinking it makes more sense to use the existing backend and rely on UrlReaders instead for all reading? That would support multiple locations from the same fetcher.

**Review by jhaals — CHANGES_REQUESTED** [review]

File: plugins/adr/src/hooks/adrFileFetcher.ts
Comment (jtreher) [reply plugins/adr/src/hooks/adrFileFetcher.ts:95]: @jhaals I was working with Robert on this and suggested that we leave the current method in place because I had seen other people mention concerns about rate limiting and a preference for using front end clients that wouldn't count against the service limit. 

I think we are all in agreement that supporting two methods drastically increases the complexity of this plugin, and it would be preferable to just use the URLReaders method as an abstraction for whatever SCM is in use. If that's the recommendation, I say we rip out Octokit.

Comment (Robunnin) [reply plugins/adr/src/hooks/adrFileFetcher.ts:95]: I believe the GitHub api is able to figure out whether or not the provided url is to a file or directory and will return either an array of files or the contents of the file accordingly (https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content). This is why beforehand it was able to have just one- EntityAdrContent was always providing links to directories and AdrReader was always providing links to files. 

With UrlReaders, there are separate methods for reading a file and reading a directory: readUrl and readTree (https://backstage.io/docs/reference/backend-common.urlreader). This is why I have two methods.

With the UrlReaders, is there a generic and reliable way to figure out if a url is to a file or directory for all the possible sites it could be used with?

File: plugins/adr/src/components/EntityAdrContent/EntityAdrContent.tsx
Comment (Robunnin) [reply plugins/adr/src/components/EntityAdrContent/EntityAdrContent.tsx:152]: I do indeed use UrlReaders for urls from multiple sites:  https://github.com/backstage/backstage/blob/0b1c9e4d651ed7267849dd08c565e82fbbc26912/plugins/adr-backend/src/service/router.ts#L34

I have expanded the adr backend plugin to provide two endpoints- one for reading a file at a provided url and another for reading a tree at a provided url. 

If you set the optional adrFileFetcher prop in EntityAdrContent to urlReaderAdrFileFetcher then it will switch to using UrlReaders. I have made it an option to use octokit or urlreaders in case people still want to use it the octokit way (for instance, if rate limiting was an issue). If that prop is not specified it will use octokit.

**Review by kuangp — CHANGES_REQUESTED** [review]
Thanks for taking a stab at implementing this! 

I left some feedback in the review comments with suggestions to modify the implementation to be more robust. Personally I think its ok to replace the octokit stuff completely and simply forward all requests to the backend to let the `UrlReaders` handle fetching adr contents. As for the rate limit concerns, I left a suggestion on how we can make use of etags and the backend plugin cache manager to reduce such a risk. Additionally, if we define a overridable api ref in the frontend, this will still allow consumers to provide their own "fetching implementation" if they want to continue making frontend requests. Again, I recommend taking a look at the TODO plugin implementation for inspiration. 

Let me know if you have any questions!

File: plugins/adr-backend/package.json
Comment (Robunnin) [reply plugins/adr-backend/package.json]: @kuangp I have switched the ```@types/express``` to be a dev dependency, however it seems the linter is now telling me to switch it back (https://github.com/backstage/backstage/actions/runs/3788464841/jobs/6441277500#step:15:11).

Is there a reason it would be doing this?

**Comment by Robunnin** [comment]
> Thanks for taking a stab at implementing this!
> 
> I left some feedback in the review comments with suggestions to modify the implementation to be more robust. Personally I think its ok to replace the octokit stuff completely and simply forward all requests to the backend to let the `UrlReaders` handle fetching adr contents. As for the rate limit concerns, I left a suggestion on how we can make use of etags and the backend plugin cache manager to reduce such a risk. Additionally, if we define a overridable api ref in the frontend, this will still allow consumers to provide their own "fetching implementation" if they want to continue making frontend requests. Again, I recommend taking a look at the TODO plugin implementation for inspiration.
> 
> Let me know if you have any questions!

@kuangp Thank you for the review and feedback! I have pushed changes that address most of your feedback. There is just one issue I am having with the dev dependency (see my other comment).

The octokit and AdrFileFetcher stuff has been removed in favor of an AdrApi. The AdrApi has caching via etags to help with rate limiting. I have updated the documentation and READMEs with information on the new implementation. I have also adjusted my version bump and added a breaking changes warning since we're now going with the UrlReaders implementation fully. 

If you get a chance and would like to, please let me know what you think of my changes and if there is anything else you would like me to change 👍

File: plugins/adr/package.json
Comment (kuangp) [inline plugins/adr/package.json:54]: looks like these new deps are no longer necessary with the latest changes?

**Review by kuangp — CHANGES_REQUESTED** [review]
One more suggested change but overall looks great to me! Great work!

File: plugins/adr-backend/package.json
Comment (kuangp) [reply plugins/adr-backend/package.json]: ah my apologies, it looks like since we are [exporting a type from the `express` package](https://github.com/backstage/backstage/pull/15348/files#diff-1b1b4bc67372dfafc27ee76b22e5b66f065e746eecc7a683133968e420623a3cR18), we also need to specify `@types/express` as an explicit dependency of this plugin

ref: [lint command](https://github.com/backstage/backstage/blob/master/packages/repo-tools/src/commands/type-deps/type-deps.ts), [reasoning](https://stackoverflow.com/questions/45176661/how-do-i-decide-whether-types-goes-into-dependencies-or-devdependencies)

Comment (Robunnin) [reply plugins/adr-backend/package.json:39]: That makes sense. 

I have adjusted the dependencies back and CI is now passing

File: plugins/adr/package.json
Comment (Robunnin) [reply plugins/adr/package.json]: I have removed these three dependencies.

**Comment by Robunnin** [comment]
@kuangp I have addressed the last suggested change. 
Please let me know if there are any more changes or things I need to do 👍

File: plugins/adr/package.json
Comment (kuangp) [inline plugins/adr/package.json:48]: Is this dep being used somewhere ? I couldn't find it being imported in this plugin

**Review by kuangp — CHANGES_REQUESTED** [review]

File: plugins/adr/package.json
Comment (Robunnin) [reply plugins/adr/package.json]: @kuangp I have removed the @backstage/catalog-client dependency. I have also removed the octokit and git-url-parse dependencies as they were left over from the old implementation and not being used anymore.

**Review by kuangp — APPROVED** [review]
Looks good on my end! 

cc: @jhaals to take a look from a maintainers perspective

**Review by jhaals — CHANGES_REQUESTED** [review]
Thanks for the excellent review work @kuangp!

This looks good to me, just the yarn.lock conflict left to fix and then this should be good to go 👍

**Comment by Robunnin** [comment]
> Thanks for the excellent review work @kuangp!
> 
> This looks good to me, just the yarn.lock conflict left to fix and then this should be good to go 👍

@jhaals The merge conflict on yarn.lock has been solved 👍

**Review by jhaals — APPROVED** [review]
👏

**Pull request merged**
